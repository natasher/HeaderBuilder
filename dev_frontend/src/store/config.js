import ui            from './modules/Common/UI'
import endpoint      from './modules/Common/Endpoint'
import uuid          from './modules/Common/UUID'
import devices       from './modules/Common/Devices'
import items         from './modules/Common/Items'
import inactiveItems from './modules/Common/InactiveItems'
import cell          from './modules/Common/Cell'
import fields        from './modules/Common/Fields'
import modals        from './modules/Common/Modals'
import general       from './modules/Common/General'
import grid          from './modules/Common/Grid'

import DesktopModule       from './modules/Desktop/'
import DesktopStickyModule from './modules/DesktopSticky'
import TabletModule        from './modules/Tablet/'
import TabletStickyModule  from './modules/TabletSticky'
import MobileModule        from './modules/Mobile/'
import MobileStickyModule  from './modules/MobileSticky'

export const storeConfig = {

  strict: true,

  modules: {
    ui,
    endpoint,
    uuid,
    devices,
    items,
    inactiveItems,
    cell,
    fields,
    modals,
    general,
    grid,

    DesktopModule,
    DesktopStickyModule,
    TabletModule,
    TabletStickyModule,
    MobileModule,
    MobileStickyModule,
  }
}