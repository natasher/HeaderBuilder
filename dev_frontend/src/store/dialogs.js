export const infoMobileInactiveItems = `In the auto mode only <b>Logo</b> and <b>Menu</b> are shown on mobile. Other items are available in <b>Custom</b> mode and are stored in Incative items below workspace.`

export const infoAutoMode = `You are going to change you header status to <b>Auto</b> mode and your work on <b>Custom</b> mode will be <u>erased</u>. Do you want to continue?`

export const onSave = `Your header has been saved.`

// export const onSaveWarning = `Lorem ipsum dolor sit amet consectetur, adipisicing elit. Neque nihil quisquam modi id, ducimus numquam.`

export const onLaunch = `Header created with the Header Builder will replace the current header.<br>If you want to use the default header managed in Theme Options, please deactivate BeTheme Header Builder plugin.`

export const onBuilderReset = `You attempt to reset your builder. All items, grid and rows data will be erased. To commit changes to WordPress click on Save changes button.`

export const afterBuilderReset = `Your Header Builder data has been erased but not saved.`