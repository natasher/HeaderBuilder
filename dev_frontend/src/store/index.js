import Vue     from 'vue'
import Vuex    from 'vuex'
import { storeConfig } from './config'

Vue.config.devtools      = true
Vue.config.productionTip = true

Vue.use( Vuex )

const store = new Vuex.Store( storeConfig )

export default store