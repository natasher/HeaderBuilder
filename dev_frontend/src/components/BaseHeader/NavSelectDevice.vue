<template>
  <nav >
    <ul class = "tabs select-device">

      <li @click = "changeToDesktop"
          :class = "{
            'active': ( getCurrentView === 'desktop' ) ? true : false
          }">
        <a data-type = "desktop">
          <span class = "icon">
            <font-icon icon = "monitor" />
          </span>
          <span class = "text">Desktop</span>
        </a>
      </li>

      <li @click = "changeToTablet"
          :class = "{
            'active': ( getCurrentView === 'tablet' ) ? true : false
          }">
        <a data-type = "tablet">
          <span class = "icon">
            <font-icon icon = "mobile-line" />
          </span>
          <span class = "text">Tablet</span>
        </a>
      </li>

      <li @click = "changeToMobile"
          :class = "{
            'active': ( getCurrentView === 'mobile' ) ? true : false
          }">
        <a data-type = "mobile">
          <span class = "icon">
            <font-icon icon = "mobile" />
          </span>
          <span class = "text">Mobile</span>
        </a>
      </li>

    </ul>
  </nav>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex'
import FontIcon from '../BaseGeneric/FontIcon.vue'
import { infoMobileInactiveItems } from '../../store/dialogs'

export default {

  name: 'NavSelectDevice',

  components: {
    FontIcon,
  },

  computed: {
    ...mapState( 'general', {
      getCurrentView: state => state.currentView,
    }),
    ...mapGetters( 'ui', [ 'getGridStatus' ]),
  },

  methods: {
    ...mapActions( 'grid', [
      'clone',
    ]),

    ...mapActions( 'inactiveItems', [
      'clear',
    ]),

    ...mapActions( 'ui', [
      'changeDevice',
    ]),

    changeToDesktop: function() {
      this.changeDevice({ device: 'desktop' })
    },

    changeToTablet: function() {
      this.changeDevice({ device: 'tablet' })
        .then(() => {
          this.$nextTick(() => {

            if (this.getGridStatus === 'auto') {
              this.clone({ from: 'DesktopModule' })
            }

          })
        })
    },

    changeToMobile: function() {
      this.changeDevice({ device: 'mobile' })
        .then(() => {
          this.$nextTick(() => {

            if (this.getGridStatus === 'auto') {
              this.clear()
              this.clone({ from: 'DesktopModule' })

              this.$modal.show('MfnGenericModal', {
                title       : 'Info',
                text        : infoMobileInactiveItems,
                closeOnClick: true,
              })
            }

          })
        })
    },

  },
}
</script>
