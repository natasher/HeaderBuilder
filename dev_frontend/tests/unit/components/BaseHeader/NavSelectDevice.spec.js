import { shallowMount, createLocalVue } from '@vue/test-utils'
import { infoMobileInactiveItems } from '../../../../src/store/dialogs'
import Vuex            from 'vuex'
import flushPromises   from 'flush-promises'
import merge           from 'lodash.merge'
import NavSelectDevice from '../../../../src/components/BaseHeader/NavSelectDevice.vue'

const localVue = createLocalVue()
localVue.use( Vuex )

function createStore(overrides) {
  const defaultStoreConfig = {
    modules: {
      general: {
        namespaced: true,
        state: {
          currentView: 'desktop'
        }
      },
      ui: {
        namespaced: true,
        getters: {
          getGridStatus: jest.fn().mockReturnValue('custom')
        },
        actions: {
          changeDevice: jest.fn()
        }
      },
      grid: {
        namespaced: true,
        actions: {
          clone: jest.fn()
        }
      },
      inactiveItems: {
        namespaced: true,
        actions: {
          clear: jest.fn()
        }
      }
    }
  }

  return new Vuex.Store(
    merge( defaultStoreConfig, overrides )
  )
}

function createWrapper(overrides) {
  const defaultMountingOptions = {
    localVue,
    store: createStore(),
    mocks: {
      $modal: {
        show: jest.fn()
      }
    }
  }

  return shallowMount( NavSelectDevice, merge( defaultMountingOptions, overrides ))
}

describe('NavSelectDevice.vue', () => {

  describe('snapshots:', () => {

    test('should match if `desktop` device is selected', () => {
      const wrapper = createWrapper()

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if `tablet` device is selected', () => {
      const store = createStore({
        modules: {
          general: {
            state: {
              currentView: 'tablet'
            }
          },
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

    test('should match if `mobile` device is selected', () => {
      const store = createStore({
        modules: {
          general: {
            state: {
              currentView: 'mobile'
            }
          },
        }
      })
      const wrapper = createWrapper({ store, })

      expect( wrapper.element ).toMatchSnapshot()
    })

  })

  test('should change device to `desktop` if desktop icon is clicked', async () => {
    const changeDevice = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            changeDevice,
          }
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'desktop'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect( changeDevice ).toBeCalledWith( expect.anything(), { device }, void 0 )
  })

  test('should change device to `tablet` if tablet icon is clicked', async () => {
    const changeDevice = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            changeDevice,
          }
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'tablet'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect( changeDevice ).toBeCalledWith( expect.anything(), { device }, void 0 )
  })

  test('should NOT clone desktop grid if tablet grid is NOT set to `auto`', async () => {
    const clone = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          getters: {
            getGridStatus: jest.fn().mockReturnValue('custom')
          },
        },
        grid: {
          actions: {
            clone
          }
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'tablet'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect( clone ).not.toBeCalled()
  })

  test('should clone desktop grid if tablet grid is set to `auto`', async () => {
    const clone = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          getters: {
            getGridStatus: jest.fn().mockReturnValue('auto')
          },
        },
        grid: {
          actions: {
            clone
          }
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'tablet'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect(clone).toBeCalledWith(expect.anything(), { from: 'DesktopModule' }, void 0)
  })

  test('should change device to `mobile` if mobile icon is clicked', async () => {
    const changeDevice = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          actions: {
            changeDevice,
          }
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'mobile'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect( changeDevice ).toBeCalledWith( expect.anything(), { device }, void 0 )
  })

  test('should NOT clone desktop grid if mobile grid is NOT set to `auto`', async () => {
    const clone = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          getters: {
            getGridStatus: jest.fn().mockReturnValue('custom')
          },
        },
        grid: {
          actions: {
            clone
          }
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'mobile'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect( clone ).not.toBeCalled()
  })

  test('should clear inactive items then clone desktop grid if mobile grid is set to `auto`', async () => {
    const clone = jest.fn()
    const clear = jest.fn()
    const store = createStore({
      modules: {
        ui: {
          getters: {
            getGridStatus: jest.fn().mockReturnValue('auto')
          },
        },
        grid: {
          actions: {
            clone,
          }
        },
        inactiveItems: {
          actions: {
            clear,
          }
        }
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'mobile'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect(clear).toBeCalled()
    expect(clone).toBeCalledWith(expect.anything(), { from: 'DesktopModule' }, void 0)
  })

  test('should open modal with info about inactive items when on `mobile`', async () => {
    const store = createStore({
      modules: {
        ui: {
          getters: {
            getGridStatus: jest.fn().mockReturnValue('auto')
          },
        },
      }
    })
    const wrapper = createWrapper({ store, })

    const device  = 'mobile'
    wrapper.find(`a[data-type="${ device }"`).trigger('click')
    await flushPromises()

    expect(wrapper.vm.$modal.show).toBeCalledWith(expect.anything(), {
      title       : 'Info',
      text        : infoMobileInactiveItems,
      closeOnClick: true
    })
  })

})