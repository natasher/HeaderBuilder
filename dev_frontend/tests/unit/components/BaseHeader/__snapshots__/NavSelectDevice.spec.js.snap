// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavSelectDevice.vue snapshots: should match if \`desktop\` device is selected 1`] = `
<nav>
  <ul
    class="tabs select-device"
  >
    <li
      class="active"
    >
      <a
        data-type="desktop"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="monitor"
          />
        </span>
         
        <span
          class="text"
        >
          Desktop
        </span>
      </a>
    </li>
     
    <li
      class=""
    >
      <a
        data-type="tablet"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="mobile-line"
          />
        </span>
         
        <span
          class="text"
        >
          Tablet
        </span>
      </a>
    </li>
     
    <li
      class=""
    >
      <a
        data-type="mobile"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="mobile"
          />
        </span>
         
        <span
          class="text"
        >
          Mobile
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`NavSelectDevice.vue snapshots: should match if \`mobile\` device is selected 1`] = `
<nav>
  <ul
    class="tabs select-device"
  >
    <li
      class=""
    >
      <a
        data-type="desktop"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="monitor"
          />
        </span>
         
        <span
          class="text"
        >
          Desktop
        </span>
      </a>
    </li>
     
    <li
      class=""
    >
      <a
        data-type="tablet"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="mobile-line"
          />
        </span>
         
        <span
          class="text"
        >
          Tablet
        </span>
      </a>
    </li>
     
    <li
      class="active"
    >
      <a
        data-type="mobile"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="mobile"
          />
        </span>
         
        <span
          class="text"
        >
          Mobile
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`NavSelectDevice.vue snapshots: should match if \`tablet\` device is selected 1`] = `
<nav>
  <ul
    class="tabs select-device"
  >
    <li
      class=""
    >
      <a
        data-type="desktop"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="monitor"
          />
        </span>
         
        <span
          class="text"
        >
          Desktop
        </span>
      </a>
    </li>
     
    <li
      class="active"
    >
      <a
        data-type="tablet"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="mobile-line"
          />
        </span>
         
        <span
          class="text"
        >
          Tablet
        </span>
      </a>
    </li>
     
    <li
      class=""
    >
      <a
        data-type="mobile"
      >
        <span
          class="icon"
        >
          <font-icon-stub
            icon="mobile"
          />
        </span>
         
        <span
          class="text"
        >
          Mobile
        </span>
      </a>
    </li>
  </ul>
</nav>
`;
